ci/order-crates-for-publishing.py

```diff
@@ -21,52 +21,26 @@ def load_metadata():
     return json.loads(subprocess.Popen(
         cmd, shell=True, stdout=subprocess.PIPE).communicate()[0])
 
-# Consider a situation where a crate now wants to use already existing
-# developing-oriented library code for their integration tests and benchmarks,
-# like creating malformed data or omitting signature verifications. Ideally,
-# the code should have been guarded under the special feature
-# `dev-context-only-utils` to avoid accidental misuse for production code path.
-#
-# In this case, the feature needs to be defined then activated for the crate
-# itself. To that end, the crate actually needs to depend on itself as a
-# dev-dependency with `dev-context-only-utils` activated, so that the feature
-# is conditionally activated only for integration tests and benchmarks. In this
-# way, other crates won't see the feature activated even if they normal-depend
-# on the crate.
-#
-# This self-referencing dev-dependency can be thought of a variant of
-# dev-dependency cycles and it's well supported by cargo. The only exception is
-# when publishing. In general, cyclic dev-dependency doesn't work nicely with
-# publishing: https://github.com/rust-lang/cargo/issues/4242 .
-#
-# However, there's a work around supported by cargo. Namely, it will ignore and
-# strip these cyclic dev-dependencies when publishing, if explicit version
-# isn't specified: https://github.com/rust-lang/cargo/pull/7333 (Released in
-# rust 1.40.0: https://releases.rs/docs/1.40.0/#cargo )
-#
-# This script follows the same safe discarding logic to exclude these
-# special-cased dev dependencies from its `dependency_graph` and further
-# processing.
-def is_self_dev_dep_with_dev_context_only_utils(package, dependency, wrong_self_dev_dependencies):
+# Cargo publish if fine with circular dev-dependencies if
+# they are path deps.
+def is_path_dev_dep(package, dependency, wrong_path_dev_dependencies):
     no_explicit_version = '*'
-
     is_special_cased = False
     if (dependency['kind'] == 'dev' and
         dependency['name'] == package['name'] and
```

Comment:
is there actual doctest code which is blocked by this pr? Depending on the context, i wonder this `self-dependency` check could also be relaxed or adjust the new comment.

Reply:
btw, relaxing this might involve to update the actual circular check as well, which might be bothersome. as the whole script itself is kind of some nice-to-have thing in ci and considering the unlikeliness of `crate A -> path dev dep -> crate B -> path dev dep -> crate C -> path dev dep with version -> crate A`, I think just commenting about incompleteness is enough...

Reply:
thanks for sharing the failing example.

so, it looks like this check (`dependency['name'] == package['name'] and`) should be removed?

```bash
$ git reset --hard e7673099ca7d159e603f7288a351234d5634a1fa
git revert e7673099ca7d159e603f7288a351234d5634a1fa

$ ./ci/order-crates-for-publishing.py 
+ exec cargo +1.78.0 metadata --no-deps --format-version=1
Error: Circular dependency: solana-pubkey <--> solana-pubkey
Error: Circular dependency: solana-program <--> solana-pubkey

$ git checkout kevinheavey/allow-path-dev-deps ./ci/order-crates-for-publishing.py
Updated 1 path from 4560e6bb3ef

$ ./ci/order-crates-for-publishing.py
+ exec cargo +1.78.0 metadata --no-deps --format-version=1
Error: Circular dependency: solana-program <--> solana-pubkey
```

that makes my previous comment (https://github.com/anza-xyz/agave/pull/2471/files#r1714630153) valid and 

> ... I think just commenting about incompleteness is enough...

this is my 2 cents. so, just add something like this below:

```python
# Check for direct circular dependencies.
# Note that this check isn't exhaustive because path dependencies can be excluded.
circular_dependencies = set()
for package, dependencies in dependency_graph.items():
```

btw, this is the reason i made the original checks overly restricted, barely white-listing the specific dcou use.

Reply:
well, i actually needed code changes like this, not just with removing the self-dep check...

```patch
$ git diff
diff --git a/ci/order-crates-for-publishing.py b/ci/order-crates-for-publishing.py
index fd763472934..a478a35200a 100755
--- a/ci/order-crates-for-publishing.py
+++ b/ci/order-crates-for-publishing.py
@@ -30,10 +30,9 @@ def is_path_dev_dep(package, dependency, wrong_path_dev_dependencies):
     no_explicit_version = '*'
     is_special_cased = False
     if (dependency['kind'] == 'dev' and
-        dependency['name'] == package['name'] and
         'path' in dependency):
         is_special_cased = True
-        if dependency['req'] != no_explicit_version:
+        # move the long dcou bla bla comment here
+        if dependency['req'] != no_explicit_version and dependency['name'] == package['name'] and 'dev-context-only-utils' in dependency['features']:
             # it's likely `{ workspace = true, ... }` is used, which implicitly pulls the
             # version in...
             wrong_path_dev_dependencies.append(dependency)

```

so, moving the removed comment and restoring the variable name `wrong_self_dev_dependencies` might be needed?